sql/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java

```diff
@@ -217,28 +222,45 @@ private static void collectFailingItems(ShardedRequests<ShardUpsertRequest, Shar
         return listener;
     }
 
-    private boolean shouldPause(ShardedRequests<ShardUpsertRequest, ShardUpsertRequest.Item> requests) {
-        if (createPartitionsRequestOngoing) {
-            LOGGER.debug("partition creation in progress, will pause");
-            return true;
-        }
-
+    private boolean shouldPauseOnTargetNodeJobsCounter(ShardedRequests<ShardUpsertRequest, ShardUpsertRequest.Item> requests) {
         for (ShardLocation shardLocation : requests.itemsByShard.keySet()) {
             String requestNodeId = shardLocation.nodeId;
             if (nodeJobsCounter.getInProgressJobsForNode(requestNodeId) >= MAX_NODE_CONCURRENT_OPERATIONS) {
-                LOGGER.debug("reached maximum concurrent operations for node {}", requestNodeId);
+                if (isDebugEnabled) {
+                    LOGGER.debug("reached maximum concurrent operations for node {}", requestNodeId);
+                }
                 return true;
             }
         }
         return false;
     }
 
+    /** @noinspection unused*/
+    private boolean shouldPauseOnPartitionCreation(ShardedRequests<ShardUpsertRequest, ShardUpsertRequest.Item> ignore) {
+        if (createPartitionsRequestOngoing) {
+            if (isDebugEnabled) {
+                LOGGER.debug("partition creation in progress, will pause");
+            }
+            return true;
+        }
+        return false;
+    }
+
     @Override
     public CompletableFuture<? extends Iterable<Row>> apply(BatchIterator<Row> batchIterator) {
         BatchIterator<ShardedRequests<ShardUpsertRequest, ShardUpsertRequest.Item>> reqBatchIterator =
             BatchIterators.partition(batchIterator, bulkSize, () -> new ShardedRequests<>(requestFactory), grouper,
                 bulkShardCreationLimiter);
 
+        // If IO is involved the source iterator should pause when the target node reaches a concurrent job counter limit.
+        // Without IO, we assume that the source iterates over in-memory structures which should be processed as
+        // fast as possible to free resources.
+        Predicate<ShardedRequests<ShardUpsertRequest, ShardUpsertRequest.Item>> shouldPause =
+            this::shouldPauseOnPartitionCreation;
+        if (batchIterator.involvesIO()) {
+            shouldPause.and(this::shouldPauseOnTargetNodeJobsCounter);
```

Comment:
doesn't the result of `and` have to be assigned to `shouldPause`? The `Predicate` isn't mutated afaik.